tests/integration/upgrade/test_sharding_upgrade.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import time
+
+import pytest
+from pytest_operator.plugin import OpsTest
+
+from ..ha_tests import helpers as ha_helpers
+from ..helpers import find_unit, unit_hostname
+from ..sharding_tests.helpers import deploy_cluster_components, integrate_cluster
+from ..sharding_tests.writes_helpers import (
+    SHARD_ONE_DB_NAME,
+    SHARD_TWO_DB_NAME,
+    count_shard_writes,
+    stop_continous_writes,
+)
+
+MONGOD_SERVICE = "snap.charmed-mongodb.mongod.service"
+MONGOS_SERVICE = "snap.charmed-mongodb.mongos.service"
+SHARD_ONE_APP_NAME = "shard-one"
+SHARD_TWO_APP_NAME = "shard-two"
+CONFIG_SERVER_APP_NAME = "config-server"
+SHARD_COMPONENTS = [SHARD_ONE_APP_NAME, SHARD_TWO_APP_NAME]
+CLUSTER_COMPONENTS = [SHARD_ONE_APP_NAME, SHARD_TWO_APP_NAME, CONFIG_SERVER_APP_NAME]
+SHARD_REL_NAME = "sharding"
+CONFIG_SERVER_REL_NAME = "config-server"
+TIMEOUT = 15 * 60
+MEDIAN_REELECTION_TIME = 12
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest) -> None:
+    """Build deploy, and integrate, a sharded cluster.
+
+    TODO: When upgrades are supported, deploy with most recent revision (6/stable when possible,
+    but 6/edge as soon as available)
+    """
+    num_units_cluster_config = {
+        CONFIG_SERVER_APP_NAME: 2,
+        SHARD_ONE_APP_NAME: 2,
+        SHARD_TWO_APP_NAME: 1,
+    }
+    await deploy_cluster_components(ops_test, num_units_cluster_config)
+
+    await ops_test.model.wait_for_idle(
+        apps=CLUSTER_COMPONENTS, idle_period=20, timeout=TIMEOUT, raise_on_blocked=False
+    )
+    await integrate_cluster(ops_test)
+    await ops_test.model.wait_for_idle(
+        apps=CLUSTER_COMPONENTS, status="active", idle_period=20, timeout=TIMEOUT
+    )
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_upgrade(
```

Comment:
We should also have a test to try the "rollback" in the middle of the upgrade.

Reply:
100% agree. Unfortunately we cannot automate this test until the charm is published on charmhub. I wrote a version of the tests that we can update once this happens